### DISCUSSION TITLE
posts/how-tty-system-works/

### DISCUSSION THREAD

**Discussion by giscus** [waynerv/waynerv.github.io#28] [Announcements]
# posts/how-tty-system-works/

长久以来，我们把终端和 shell 视作开发者理所当然的基本工具，但其底层涉及的组件和机制却并不简单。本文将帮助你理清终端、TTY、伪终端、PTY 和 Shell 等概念。

https://waynerv.com/posts/how-tty-system-works/

**Comment by yurenchen000** [comment]
👍博客很有质感。

文章一旦长了，滚回目录不太容易
目录要是能随时呼出就好了。

**Reply by yurenchen000** [reply]
像是黑夜找东西似的 一阵四处摸索 🙈
  没有找到呼出 `悬浮TOC` 的机关

**Reply by waynerv** [reply]
因为目前确实没有 😅

**Reply by markity** [reply]
太不错了

**Comment by markity** [comment]
请问终端改变窗口大小怎么传递给服务端呢

**Reply by yurenchen000** [reply]
如果 ssh 的话, 记得它有自己的报文传递 win size change 事件 (ssh client 拿到win size 同步给 ssh server 的 pts 逻辑尺寸).
另外 终端里运行的程序 也可以通过转义序列 从终端查询(需要终端软件的支持) 终端窗口大小.

转义序列有没有事件不记得了,应该是有. 在配置terminal mode 的时候 配置 终端软件 发送 win size change 之类的事件, 也是通过转义序列.

https://terminalguide.namepad.de/seq/
你可以参考这里, 还有 xterm 的一些协议文档.
`man console_codes` 里面的很不全.

**Reply by yurenchen000** [reply]
物理 tty 的 size 一般则是固定的分辨率(和屏幕分辨率类似), 不会在运行时变化.
(它可以通过 ioctl 从 kernel 拿 tty 尺寸, 好像也能改).

ssh (及 pts) 这类 是 CS 架构.
tty 则连界面(屏幕或串口) 都是 kernel 在维持.

**Comment by markity** [comment]
嗯, 我单独设计了下数据包，把winsize改变单独通知给服务器，服务器收到这个数据包后设置伪终端的winsize, kernel就会给伪终端的前台程序发WINCH signal, 这样行为就正常了

**Reply by markity** [reply]
cooked mode下行为还是和ssh的显示有区别
比如`nc`命令(netcat)输入中文，然后按backspace, 它们行为是不一样的，不知道这是什么原因

**Reply by markity** [reply]
补充下, 在建立连接时也需要先传递window size, pty初始化时用这个传递过去的window size

**Comment by liwenkun** [comment]
写得很好👍🏻。我有一个问题请教一下作者，如今的 shell 和终端都是软件，因此终端启动可以带着 shell 也启动，这没啥问题。但以前的 shell 和终端是分开的，shell 是不是必须得开机启动，并常驻后台去监听来自终端的输入输出？还是说连接上终端后，系统会自动启动一个 shell 进程去负责读取来自这个终端的输入输出呢？

**Reply by yurenchen000** [reply]
tty shell 貌似是 agetty 或其他软件 (相当于 sshd 的角色) 来接收请求(从键盘或串口), 再起 shell.

 更早的 终端硬件 我就不知道了.

**Comment by noname007** [comment]
内容写的很棒！👍，看到伪终端的时候突然解决了我以前一直没想明白的问题（标准输入在不同进程间怎么区分开的）：
1. 写一个从stdin 读数据的程序
2. iterm 中开两个标签页面，每个页面都运行程序 1 中的程序
3. 键盘输入的数据，会被哪个进程读取？
